### PR TITLE
feat(parameters): accept boto3_client to support private endpoints and ease testing

### PR DESCRIPTION
**Pull request by ran-isenberg**
**Fixes:** #1079

### Changes

> Please provide a summary of what's being changed

This adds a new `boto3_client` constructor parameter to the Parameters feature to unblock customers using [VPC endpoints](https://docs.aws.amazon.com/vpc/latest/privatelink/vpc-endpoints.html).

~~I didn't add it to DynamoDB parameter because i'm not sure it's possible there. Fixed also the docs.~~ All clients have it now, and DynamoDB has a note in docstrings + type referring to Resource Client (high-level) not Client (low-level) like SSM, SecretsManager and AppConfig.

### User experience

> Please share what the user experience looks like before and after this change

```python
import boto3

from aws_lambda_powertools.utilities import parameters

# Initialize clients with custom VPC endpoint
ssm = boto3.client("ssm", endpoint_url="custom_endpoint")
secrets = boto3.client("secretsmanager", endpoint_url="custom_endpoint")
appconfig = boto3.client("appconfig", endpoint_url="custom_endpoint")


ssm_provider = parameters.SSMProvider(boto3_client=ssm)
secrets_provider = parameters.SecretsProvider(boto3_client=secrets)
appconf_provider = parameters.AppConfigProvider(environment="my_env", application="my_app", boto3_client=appconfig)
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of my this change
* [x] Changes are tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

-----
[View rendered docs/utilities/parameters.md](https://github.com/ran-isenberg/aws-lambda-powertools-python/blob/boto/docs/utilities/parameters.md)

### CODE REVIEW

**Comment by ran-isenberg** [comment]
@heitorlessa  not really sure how to add tests for the client init, didnt see any tests for the session either

**Comment by ran-isenberg** [comment]
not really sure why codecov doesnt show an updated version. I added tests

**Comment by heitorlessa** [comment]
thanks for that @ran-isenberg - fixed the title as I see this as a bug as customers can't use the functionality we provide. 

The pull request body is missing our [required acknowledgements](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/PULL_REQUEST_TEMPLATE.md): are you okay if I edit the content to add them back as I can't review PRs without it?

Thanks a lot!

**Comment by ran-isenberg** [comment]
@heitorlessa  after all this time, do you really need to ask? :)
go for it

File: docs/utilities/parameters.md
Comment (heitorlessa) [inline docs/utilities/parameters.md:485]:
```suggestion
The **`config`** , **`boto3_session`**, and **`boto3_client`**  parameters enable you to pass in a custom [botocore config object](https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html) , [boto3 session](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html), or  a [boto3 client](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/boto3.html) when constructing any of the built-in provider classes.
```

File: aws_lambda_powertools/utilities/parameters/ssm.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parameters/ssm.py:24]: Let's use SSMClient from mypy boto3 project as it accurately defines what we're expecting, so mypy and other type checkers can catch potential issues later.

We can bring it as a dev dependency, and use `"SSMClient"` to benefit from Python's forward reference type annotation.

https://youtype.github.io/boto3_stubs_docs/mypy_boto3_ssm/client/

File: aws_lambda_powertools/utilities/parameters/secrets.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parameters/secrets.py:69]: Same as the SSM comment: https://youtype.github.io/boto3_stubs_docs/mypy_boto3_secretsmanager/#secretsmanagerclient

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parameters/appconfig.py:73]: Same as the SSM comment: https://youtype.github.io/boto3_stubs_docs/mypy_boto3_appconfig/client/

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parameters/appconfig.py:85]: Can we turn this into a builder static method in the parent class? Return type could be Any as we will be able to cast the correct one for each client built w/ mypy_boto3.

This will prevent future an accidental logic discrepancy between them and localize change. It could accept the `boto3_client`, `boto3_session`, `boto3_config`, and `boto3_service_name` to initialize.

Let me know if it doesn't make sense

**Review by heitorlessa — CHANGES_REQUESTED** [review]
Two minor changes - **1/** Make use of forward reference for type annotation ( "SSMClient" from `mypy_boto3_<service`), and **2/** refactor branching logic into a static method to prevent logic discrepancy in a future change.

I ran out of time to look into the tests and DynamoDB, but I will as soon as I'm able.

Side note: Secrets Manager and SSM seem to be the only ones supporting `endpoint_url` according to this doc, yet boto3_client will benefit everyone's testing: https://docs.aws.amazon.com/vpc/latest/privatelink/integrated-services-vpce-list.html

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parameters/appconfig.py:85]: i didnt add it to the dymamo class becuase we dont use a client there, not usre you can create a table from a client class.
so if it's not the dynamodb, it shouldn't be in the parent

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parameters/appconfig.py]: fixed

File: aws_lambda_powertools/utilities/parameters/secrets.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parameters/secrets.py]: fixed

File: aws_lambda_powertools/utilities/parameters/ssm.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parameters/ssm.py]: fixed

**Comment by michaelbrewer** [comment]
> Two minor changes - **1/** Make use of forward reference for type annotation ( "SSMClient" from `mypy_boto3_<service`), and **2/** refactor branching logic into a static method to prevent logic discrepancy in a future change.
> 
> I ran out of time to look into the tests and DynamoDB, but I will as soon as I'm able.
> 
> Side note: Secrets Manager and SSM seem to be the only ones supporting `endpoint_url` according to this doc, yet boto3_client will benefit everyone's testing: [docs.aws.amazon.com/vpc/latest/privatelink/integrated-services-vpce-list.html](https://docs.aws.amazon.com/vpc/latest/privatelink/integrated-services-vpce-list.html)

@heitorlessa @ran-isenberg - should we be bundling `boto3-stubs` as part of the python runtime?

**Comment by heitorlessa** [comment]
Haven't forgotten @ran-isenberg. I'll look into as soon as I can this week.

**Comment by michaelbrewer** [comment]
When this is released as is and it has to include a new runtime dependency for internal typing. 

Can we include a warning in the changelog for customers like me at FiServ, who needs to scan for upstream issues and resolve conflicts if using a conflicting boto stubs. 

Or change this make this optional dependency

**Comment by michaelbrewer** [comment]
maybe 2 PRs 
- 154 lines to fix just #1079
- 430 lines or more for the mypy stubs and what ever other packages could use it. Like Idempotency, Batch, Parameters, Data Class (`code_pipeline_job_event`)

**Review by heitorlessa — CHANGES_REQUESTED** [review]
@ran-isenberg thanks for making those changes. I've made code change suggestions along with explanations about why they're needed - let me know if they don't make sense.

For the DynamoDB and parent static method piece, GitHub seems to have swallowed the review thread and I can't make comments anymore. I understand you concern about having a static method in the parent that's applicable to two clients (SSM, SecretsManager) but  DynamoDB. I'd like to propose a solution to unblock this PR, as we still need to address DynamoDB regardless.

**Suggestion**: Add the same `boto3_client` in DynamoDB with an explicit resource type definition, **plus** doc strings as well as docs stating that we expect a `DynamoDB resource client`. I thought about `boto3_resource_client` but ultimately it will come down to practicality at this point, hence the suggestion of keeping `boto3_client`, as resource is also a client (high level vs low level).

**Comment by michaelbrewer** [comment]
> @ran-isenberg thanks for making those changes. I've made code change suggestions along with explanations about why they're needed - let me know if they don't make sense.
> 
> For the DynamoDB and parent static method piece, GitHub seems to have swallowed the review thread and I can't make comments anymore. I understand you concern about having a static method in the parent that's applicable to two clients (SSM, SecretsManager) but DynamoDB. I'd like to propose a solution to unblock this PR, as we still need to address DynamoDB regardless.
> 
> **Suggestion**: Add the same `boto3_client` in DynamoDB with an explicit resource type definition, **plus** doc strings as well as docs stating that we expect a `DynamoDB resource client`. I thought about `boto3_resource_client` but ultimately it will come down to practicality at this point, hence the suggestion of keeping `boto3_client`, as resource is also a client (high level vs low level).

Great work in finding an alternative of doing typing checking during development time and not make a required runtime dependency. 

But does this make sense as part of this pull request? At a minimum just the added client parameter is needed to resolve #1079

File: docs/utilities/parameters.md
Comment (ran-isenberg) [reply docs/utilities/parameters.md:27]: yes I am. markdown editor

File: aws_lambda_powertools/utilities/parameters/secrets.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parameters/secrets.py:70]: fixed

File: aws_lambda_powertools/utilities/parameters/secrets.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parameters/secrets.py:10]: fixed

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parameters/appconfig.py:12]: fixed

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (ran-isenberg) [reply aws_lambda_powertools/utilities/parameters/appconfig.py:74]: fixed

**Comment by ran-isenberg** [comment]
@heitorlessa  @michaelbrewer  I pushed the changes you wanted.
I didnt do the dynamodb changes as I dont fully understand what you suggested. DynamoDB already gets an optional boto session, so not really sure what the client has to do with anything. We create a table from the session.

Feel free to make the changes yourself or explain them again :)

**Comment by heitorlessa** [comment]
> @heitorlessa @michaelbrewer I pushed the changes you wanted. I didnt do the dynamodb changes as I dont fully understand what you suggested. DynamoDB already gets an optional boto session, so not really sure what the client has to do with anything. We create a table from the session.
> 
> Feel free to make the changes yourself or explain them again :)

Thanks a lot for making those changes, Ran! I'll push the other changes and ping you for review later this week

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/parameters/appconfig.py:87]:
```suggestion
            self.client: "AppConfigClient" = session.client("appconfig", config=config)
```

File: aws_lambda_powertools/utilities/parameters/secrets.py
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/parameters/secrets.py:83]:
```suggestion
            self.client: "SecretsManagerClient" = session.client("secretsmanager", config=config)
```

File: aws_lambda_powertools/utilities/parameters/ssm.py
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/parameters/ssm.py:99]:
```suggestion
            self.client: "SSMClient" = session.client("ssm", config=config)
```

**Review by michaelbrewer — CHANGES_REQUESTED** [review]
I would wrap quotes around all of the type hints.

File: pyproject.toml
Comment (michaelbrewer) [inline pyproject.toml:59]: @ran-isenberg @heitorlessa - i do wonder how we can run the unit tests without `mypy-boto3-*` being installed. Then we can be certain the missing imports don't cause runtime errors. Maybe we should make a note of this?

later integration test would help solve this gap.

**Review by michaelbrewer — APPROVED** [review]
LGTM, other than the gaps on the testing side.

File: docs/utilities/parameters.md
Comment (michaelbrewer) [inline docs/utilities/parameters.md:531]: I would update the example to still use boto session to create the client  (see [boto3 Sessions, and Why You Should Use Them](https://ben11kehoe.medium.com/boto3-sessions-and-why-you-should-use-them-9b094eb5ca8e) ). And showcase what you can do when you create your own client, like customerize the endpoint url as per the original issue.

```suggestion
	from aws_lambda_powertools.utilities import parameters
	import boto3

        session = boto3.session.Session()
	boto3_client= session.client(service_name="ssm", endpoint_url='custom_endpoint')
	ssm_provider = parameters.SSMProvider(boto3_client=boto3_client)

	def handler(event, context):
		# Retrieve a single parameter
		value = ssm_provider.get("/my/parameter")
		...
```

**Comment by michaelbrewer** [comment]
@ran-isenberg @heitorlessa - in case you want a dumbed down version of this PR see #1197 . It just focuses on the original request and docs updates. This might make review easier. 

> feel free to kill #1197 if it does not fit

**Comment by heitorlessa** [comment]
> > @heitorlessa @michaelbrewer I pushed the changes you wanted. I didnt do the dynamodb changes as I dont fully understand what you suggested. DynamoDB already gets an optional boto session, so not really sure what the client has to do with anything. We create a table from the session.
> > Feel free to make the changes yourself or explain them again :)
> 
> Thanks a lot for making those changes, Ran! I'll push the other changes and ping you for review later this week

As promised, I've transferred what I had in my local branch and made additional fixes. This is now ready to be review.

@gshpychka - @sthulb will review and once approved we should release it ASAP to unblock the VPC private endpoints. We had a delay in getting this out to you due to a larger operational and internal work.

**Review by heitorlessa — APPROVED** [review]

**Comment by michaelbrewer** [comment]
> > > @heitorlessa @michaelbrewer I pushed the changes you wanted. I didnt do the dynamodb changes as I dont fully understand what you suggested. DynamoDB already gets an optional boto session, so not really sure what the client has to do with anything. We create a table from the session.
> > > Feel free to make the changes yourself or explain them again :)
> > 
> > 
> > Thanks a lot for making those changes, Ran! I'll push the other changes and ping you for review later this week
> 
> As promised, I've transferred what I had in my local branch and made additional fixes. This is now ready to be review.
> 
> @gshpychka - @sthulb will review and once approved we should release it ASAP to unblock the VPC private endpoints. We had a delay in getting this out to you due to a larger operational and internal work.

Would this be a major release update? Have you done integration testing (due to the large number of changes)

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/parameters/appconfig.py:37]: that would be a breaking change.

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/parameters/appconfig.py:82]: Why is this being called twice ? 

`super().__init__()`

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/parameters/appconfig.py:37]: when a customer uses positional arguments, changing the order would be a breaking change

**Comment by michaelbrewer** [comment]
> > Would this be a major release update? Have you done integration testing (due to the large number of changes)
> 
> As per semver, this would require a minor release. 1.x.0, this doesn't remove any functionality or fundamentally change anything for users.

so a feature release ie: `1.25.0`. I never expect this to be a `2.0.0` release.

File: docs/utilities/parameters.md
Comment (michaelbrewer) [reply docs/utilities/parameters.md:27]: for #1197  i did not include this and kept the changes within scope and with less than 1/2 the lines of code changes.

File: aws_lambda_powertools/utilities/parameters/appconfig.py
Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/parameters/appconfig.py:82]: @heitorlessa i assume you meant to delete line 94 rather than call it twice?

File: aws_lambda_powertools/utilities/parameters/dynamodb.py
Comment (heitorlessa) [reply aws_lambda_powertools/utilities/parameters/dynamodb.py:161]: Agreed on the docs, I will make those changes tomorrow (docs, not param name) and merge it.

Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/parameters/dynamodb.py:161]: thanks clarity always helps.

File: aws_lambda_powertools/utilities/parameters/dynamodb.py
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/parameters/dynamodb.py:166]:
```suggestion
        client: "DynamoDBServiceResource" = self._build_boto3_resource_client(
```

There is no need to add `client` to self, and note `self.client` would make it part of the public api.

And only `self.table` benefits from typing, as it is used further down

File: aws_lambda_powertools/utilities/parameters/dynamodb.py
Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/parameters/dynamodb.py:161]: overall for `DynamoDBProvider` the resource client so that you can set the endpoint url seems redundant when you already have `endpoint_url` as a parameter. 

With 8 parameters now, this is borderline code smell.

File: docs/utilities/parameters.md
Comment (ran-isenberg) [reply docs/utilities/parameters.md:27]: > @ran-isenberg what plugin is doing this for you?

I believe it's Yapf

Comment (ran-isenberg) [reply docs/utilities/parameters.md:27]: > @ran-isenberg what plugin is doing this for you?

Markdown all in one
https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one

File: aws_lambda_powertools/utilities/parameters/dynamodb.py
Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/parameters/dynamodb.py:161]: @heitorlessa  the docs definitely helps! It might be a good general call out to recommend people use boto3 session where possible

**Pull request merged**
